Automattic/woocommerce-payments
# Display payment error message in the Payment context with Blocks

Fixes https://github.com/Automattic/woocommerce-payments/issues/7691

#### Changes proposed in this Pull Request
This PR modifies the error message location in Blocks checkout, ensuring it appears within the context of the payment elements. Previously, error messages were always displayed at the top of the page with WooCommerce Blocks. This issue arose because, during a payment processing error, an exception was thrown by WooPayments, causing WooCommerce to exit the execution prematurely. As a result, it could not properly handle the errors or emit checkout failure events to manage the error appropriately.


Instead of throwing an error, we now add the error notice (to avoid regressions with shortcode) and then include the payment method error in the payment result object. [This object is passed to us through the action hook fired by WooCommerce Core](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/src/StoreApi/Utilities/CheckoutTrait.php#L83). This approach aligns with the one used in [the Stripe Gateway implementation](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/3376adee168dbe10689fef30e2346b93e8b0461c/includes/class-wc-stripe-blocks-support.php#L381-L388).









#### Testing instructions





* Use the declined `4000000000000002` card in Blocks checkout and confirm that `Error: Your card was declined.` is displayed within payments block and not at the top of the page
* Perform the same test but in shortcode and confirm the error message didn't change (it's positioned at the top of the page, which is the expected behavior)
* Try `4000000000009995` with Blocks and confirm that the error message changed and is positioned correctly as well

-------------------


**Post merge**
